src/modules/expression-evaluator/src/resolvers/buildObject.ts

```diff
@@ -0,0 +1,35 @@
+import { EvaluateExpressionInstance, ValueNode, EvaluatorNode } from '../types'
+
+export type BuildObjectQuery = {
+  operator: 'buildObject'
+  properties: {
+    key: EvaluatorNode
+    value: EvaluatorNode
+  }[]
+}
+
+type BuildObject = (
+  buildObjectQuery: BuildObjectQuery,
+  evaluateExpressionInstance: EvaluateExpressionInstance
+) => Promise<ValueNode>
+
+const buildObject: BuildObject = async (buildObjectQuery, evaluateExpressionInstance) => {
+  const result: { [key: string]: ValueNode } = {}
+
+  const evaluateKeyValue = async (key: EvaluatorNode, value: EvaluatorNode) => {
+    const [resultKey, resultValue] = await Promise.all<ValueNode>([
+      evaluateExpressionInstance(key),
+      evaluateExpressionInstance(value),
+    ])
```

Comment:
I think I can see why you did it like this, but I think it's cleaner if each operator/resolver receives pre-evaluated "children"/"properties". If we extend later this as per your suggestions (i.e. named parameters rather than "children" array), we don't want to have to write this same "evaluate all my children" code inside each operator. Surely it's simpler to evaluate all children/properties at the top level and pass results to each operator (like it does with the current "children" array).

I think it would be fairly straightforward to extend the existing "evaluate the children" code to also iterate over additional object keys and evaluate them too, don't you think?

Reply:
I think this would also prevent the need to pass in a "evaluateExpressionInstance" -- it would all be done already.

Reply:
Sure, if there is obvious similarity in some shapes of configurations we can either add helper methods or find out a generic way of passing through configurations of evaluations that can be done by core evaluator logic. (would prefer adding helper methods, in evaluation instance that is passed through, i.e. evaluateArray(array), evaluateArrayObjects(array, ['key', 'value'])

Although I dont' think it would be that simple for some:

I guess a good way to address this point is -> if we see repetitions of evaluation in future resolvers we can consolidate them with some common logic (as long we don't sacrifice readability for the sake of dry)

Reply:
We currently iterate over each element in "children" and evaluate them before passing the children array to the relevant operator. Surely we can do the same here -- just iterate over the keys of the object (excluding "operator" and "children"), and evaluate them before passing to "buildObject". I don't think any operators need to be doing recursive calls to their nested elements when we already have functionality to handle this at the top level. It just makes it very confusing to understand to have two different ways of handling child elements.

Reply:
Just to demonstrate, I made a quick branch off this branch called `pre-evaluate-other-fields`. It's pretty bare bones and doesn't handle errors or anything, but it demonstrates how much simpler the "buildQuery" operator can (should?) be. I really think we should stick to the pattern of keeping the operators just doing their basic operation, and all the recursive calling in the main "evaluateExpression" function and not mix and match them. 

My changes:
- drastically simplify `buildObject` operator to only do its "operator" duties and nothing else
- In the same place that all "children" are resolved, wrote extra code to check and resolve expressions on any other properties of the passed-in expression (including arrays). This means we can re-use the same logic when we refactor/create other operators based on your named parameters suggestion. It's a bit ugly currently, but just trying to make the point. 

Only tested in GUI builder thus far, but working fine for basic cases.

Let me know what you think.

Reply:
Did you test in GUI after updating to use new `buildObject` ? Good way to check if it works is to run `yarn test buildObject` from evaluator module folder. 

To work for buildObject, the logic would also need to iterate over keys of objects inside the array, and try to evaluate them, and a small change in buildObject to make sure undefined values or keys are not included. I am not yet convinced that generic evaluation is worth the added complexity, unless it can be written in a readable way and all use cases are considered.

>  It just makes it very confusing to understand to have two different ways of handling child elements.

I would have thought that outsourcing logic of evaluation to resolvers simplifies things.

> It just makes it very confusing to understand to have two different ways of handling child elements.
> I don't think any operators need to be doing recursive calls to their nested elements when we already have functionality to handle this at the top.

It's quite a bit different, since the shape of configuration object is different (and may vary for future resolvers), in this case we just want to evaluate two keys (key and value) inside objects inside an array, vs evaluating children in an array. 

Anyways, I dont' think this is a priority for now, but if you come up with a readable solution, be keen to check it out. (just be wary of trying to do too much in one place to accomodate all use cases)

Reply:
Yeah, the evaluation of the `{key..., value:...}` objects is the only bit that can't easily be handled by a generic process, so it makes sense the way you've done it. Although I'm not sure I see the need for a unique format, when we could have just done:
```
{
 operator: 'buildObject',
 children: [ <key>, <value>, <key>, <value>... ]
}
```
which would have made it a lot simpler.

I get that named parameters makes sense for the operators that have a very specific, ordered set of nodes (i.e. GraphQL, API etc.), but when the nodes are all a homogeneous collection (or close to it), what's wrong with `children` array? (You could always name it "properties" instead of "children", but make the shape the same)

Anyway, all good for now, but I think we should plan for further extensions so we don't end up with every operator having its own logic and style. I think named parameters/fields is a good idea, but I think it should be possible to do it all in a generic way so as to keep the operators simple.